Failed to build for `aarch64-pc-windows-msvc`

As far as I know, wasmer can target Windows ARM64, and so do cranelift.

https://github.com/bytecodealliance/wasmtime/blob/f242975c49385edafe4f72dfa5f0ff6aae23eda3/crates/runtime/src/traphandlers/windows.rs#L54-L62

Hi, thanks for your replies!



> Windows on AArch64 has two possible abi's. The first is I believe AAPCS compatible except with SEH instead of DWARF based unwinding tables. The second is the "emulator compatible" abi, which as far as I understand it is basically the x86_64 abi except with x86_64 registers mapped onto AArch64 registers.

👍
See also the Microsoft Documentation about the Windows ARM64 ABI convention: https://learn.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions



> The latter one is ARM64EC ABI, which seems to be not supported by Rust. I suggest focusing on the normal ARM64 ABI.

Yes, ARM64 would also be the one we are interested in (.NET for example doesn't support ARM64EC).



> There's also the question of testing in CI: we have Windows/x86-64 runners, but does Windows have a "reverse emulator" that can run aarch64 `.exe`s on x86-64, like it can the other way around? If not, I suppose we could treat this the same as macOS/aarch64: support in-tree, but not tested so best-effort.

To my knowledge, Window ARM64 supports emulation for x86/64 processes, but unfortunately not the other way round. The only way I can think of to run Windows Arm64 binaries on x86-64 hardware would be using a VM with emulation, e.g. QEMU.


I would be happy to try to build wasmtime for Windows ARM64 and test it locally, to see if it basically works. Unfortunaltey I don't know Rust and am not very experienced with low-level handling. When I try to build using `cargo build --target aarch64-pc-windows-msvc`, I get the `"unsupported platform"` error as mentioned by @Berrysoft:
```
   Compiling wasmtime-environ v2.0.0 (C:\Users\Administrator\Desktop\Wasmtime-Test\wasmtime\crates\environ)
error: unsupported platform
  --> crates\runtime\src\traphandlers\windows.rs:62:17
   |
62 |                 compile_error!("unsupported platform");
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0425]: cannot find value `ip` in this scope
  --> crates\runtime\src\traphandlers\windows.rs:65:49
   |
65 |         let jmp_buf = info.take_jmp_buf_if_trap(ip, |handler| handler(exception_info));
   |                                                 ^^ not found in this scope

error[E0425]: cannot find value `ip` in this scope
  --> crates\runtime\src\traphandlers\windows.rs:71:31
   |
71 |             info.set_jit_trap(ip, fp);
   |                               ^^ not found in this scope

error[E0425]: cannot find value `fp` in this scope
  --> crates\runtime\src\traphandlers\windows.rs:71:35
   |
71 |             info.set_jit_trap(ip, fp);
   |                                   ^^ not found in this scope
```

I think the `(*(*exception_info).ContextRecord)` refers to the platform-specific [`CONTEXT`](https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-context) structure (here for x86_64), which for ARM64 is defined as alias for [`ARM64_NT_CONTEXT`](https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-arm64_nt_context). However, I'm not sure which values/registers to use from that struct for the `ip` and `fp` variables, or if there are other changes required.

Thanks!

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -184,6 +184,8 @@ jobs:
     - run: cargo check --target armv7-unknown-linux-gnueabihf -p wasi-common
 
   # Check whether `wasmtime` cross-compiles to aarch64-pc-windows-msvc
+  # We don't build nor test it because it lacks trap handling.
+  # Tracking issue: https://github.com/bytecodealliance/wasmtime/issues/4992
   checks_winarm64:
     name: Check Windows ARM64
     runs-on: windows-latest

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -192,8 +192,7 @@ jobs:
       with:
         submodules: true
     - uses: ./.github/actions/install-rust
-
-    # Check some feature combinations of the `wasmtime` crate
+    - run: rustup target add aarch64-pc-windows-msvc
     - run: cargo check -p wasmtime --target aarch64-pc-windows-msvc
 
   fuzz_targets:

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -183,6 +183,19 @@ jobs:
     - run: cargo check --target wasm32-unknown-emscripten -p wasi-common
     - run: cargo check --target armv7-unknown-linux-gnueabihf -p wasi-common
 
+  # Check whether `wasmtime` cross-compiles to aarch64-pc-windows-msvc
+  checks:
+    name: Check Windows ARM64
+    runs-on: windows-latest
+    steps:
+    - uses: actions/checkout@v2
+      with:
+        submodules: true
+    - uses: ./.github/actions/install-rust
+
+    # Check some feature combinations of the `wasmtime` crate
+    - run: cargo check -p wasmtime --target aarch64-pc-windows-msvc
+
   fuzz_targets:
     name: Fuzz Targets
     runs-on: ubuntu-latest

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -184,7 +184,7 @@ jobs:
     - run: cargo check --target armv7-unknown-linux-gnueabihf -p wasi-common
 
   # Check whether `wasmtime` cross-compiles to aarch64-pc-windows-msvc
-  checks:
+  checks_winarm64:
     name: Check Windows ARM64
     runs-on: windows-latest
     steps:

--- a/crates/jit/src/unwind.rs
+++ b/crates/jit/src/unwind.rs
@@ -5,6 +5,9 @@ cfg_if::cfg_if! {
     } else if #[cfg(all(windows, target_arch = "x86"))] {
         mod winx32;
         pub use self::winx32::*;
+    } else if #[cfg(all(windows, target_arch = "aarch64"))] {
+        mod winx64;
+        pub use self::winx64::*;
     } else if #[cfg(unix)] {
         mod systemv;
         pub use self::systemv::*;

--- a/crates/jit/src/unwind/winx64.rs
+++ b/crates/jit/src/unwind/winx64.rs
@@ -21,7 +21,7 @@ impl UnwindRegistration {
         if RtlAddFunctionTable(
             unwind_info as *mut _,
             (unwind_len / unit_len) as u32,
-            base_address as u64,
+            base_address as _,
         ) == 0
         {
             bail!("failed to register function table");

--- a/crates/runtime/src/traphandlers/windows.rs
+++ b/crates/runtime/src/traphandlers/windows.rs
@@ -58,6 +58,9 @@ unsafe extern "system" fn exception_handler(exception_info: *mut EXCEPTION_POINT
             } else if #[cfg(target_arch = "x86")] {
                 let ip = (*(*exception_info).ContextRecord).Eip as *const u8;
                 let fp = (*(*exception_info).ContextRecord).Ebp as usize;
+            } else if #[cfg(target_arch = "aarch64")] {
+                let ip = (*(*exception_info).ContextRecord).Pc as *const u8;
+                let fp = (*(*exception_info).ContextRecord).Anonymous.Anonymous.Fp as usize;
             } else {
                 compile_error!("unsupported platform");
             }

